packages/aws-cdk/lib/api/hotswap/step-functions.ts

```diff
@@ -0,0 +1,113 @@
+  logicalId: string, change: cfn_diff.ResourceDifference, assetParamsWithEnv: { [key: string]: string },
+): ChangeHotswapResult {
+  const stepDefinitionChange = isStepFunctionDefinitionOnlyChange(change, assetParamsWithEnv);
+
+  if ((stepDefinitionChange === ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT) ||
+      (stepDefinitionChange === ChangeHotswapImpact.IRRELEVANT)) {
+    return stepDefinitionChange;
+  }
+
+  let machineName: string | undefined;
+  try {
+    machineName = stringifyPotentialCfnExpression(change.newValue?.Properties?.StateMachineName, assetParamsWithEnv);
+  } catch (e) {
+    // It's possible we can't evaluate the function's name -
+    // for example, it can use a Ref to a different resource,
+    // which we wouldn't have in `assetParamsWithEnv`.
+    // That's fine though - ignore any errors,
+    // and treat this case the same way as if the name wasn't provided at all,
+    // which means it will be looked up using the listStackResources() call
+    // by the later phase (which actually does the StepFunctions state machine update)
+    machineName = undefined;
+  }
+
+  return new StepFunctionHotswapOperation({
+    logicalId: logicalId,
+    definition: stepDefinitionChange,
+    stateMachineName: machineName,
+  });
+}
+
+/**
+ * Returns `ChangeHotswapImpact.IRRELEVANT` if the change is not for a `AWS::StepFunctions::StateMachine`,
+ * but doesn't prevent short-circuiting
+ * (like a change to CDKMetadata resource),
+ * `ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT` if the change is to a AWS::StepFunctions::StateMachine,
+ * but not only to its definition property,
+ * or the definition string if the change is to a AWS::StepFunctions::StateMachine,
+ * and only affects its definition property.
+ */
+function isStepFunctionDefinitionOnlyChange(
+  change: cfn_diff.ResourceDifference, assetParamsWithEnv: { [key: string]: string },
+): string | ChangeHotswapImpact {
+  // if we see a different resource type, it will be caught by isNonHotswappableResourceChange()
+  // this also ignores Metadata changes
+  const newResourceType = change.newValue?.Type;
+  if (newResourceType !== 'AWS::StepFunctions::StateMachine') {
+    return ChangeHotswapImpact.IRRELEVANT;
+  }
+
+  if (change.oldValue?.Type == null) {
+    // can't short-circuit a brand new StateMachine
+    return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+  }
+
+  const propertyUpdates = change.propertyUpdates;
+
+  // ensure that only changes to the definition string result in a hotswap
+  for (const updatedPropName in propertyUpdates) {
+    const updatedProp = propertyUpdates[updatedPropName];
+
+    for (const newPropName in updatedProp.newValue) {
+      if (newPropName === 'Fn::Join') {
+        const joinString = updatedProp.newValue[newPropName];
+        const updatedDefinition = JSON.parse(joinString[1]);
+
+        return stringifyPotentialCfnExpression(JSON.stringify(updatedDefinition), assetParamsWithEnv);
+      }
```

Comment:
So you're saying if my `DefinitionString` is just `'{"A":"B"}'`, that can't be hotswapped?

Something is off here 😜.

Reply:
Also, this is a loop that only ever has one iteration... that's a pretty big code smell 🙂.